develop/2017-05-29-fgrehm-analysis-on-meals-based-on-presence-information.py

```diff
@@ -0,0 +1,563 @@
+    ~meals_outside_df_during_speeches.company_name.str.contains('HOTEL')
+]
+print("Meals excluding expenses on hotels:", len(meals_outside_df_during_speeches))
+
+
+# In[19]:
+
+def score(meal):
+    return score_by_time(meal) + score_by_ts(meal) - penalty_by_date(meal)
+
+def score_by_time(meal):
+    if meal.total_speeches > 1:
+        return meal.first_speech_at.hour - meal.last_speech_at.hour + (meal.total_speeches - 1)
+    else:
+        return 0
+    
+def score_by_ts(meal):
+    if meal.timestamps == '':
+        return 0
+    
+    occurences = 0
+    for ts in meal.timestamps:
+        if ts == None:
+            continue
+        if (meal.first_speech_at.hour-1) < ts.tm_hour < (meal.last_speech_at.hour+1):
+            occurences += 1
+    return occurences * 10
+
+def penalty_by_date(meal):
+    if meal.receipt_dates == '':
+        return 0
+    
+    bad_occurences = 0
+    for dt in meal.receipt_dates:
+        if dt == None:
+            continue
+        if dt != meal.issue_date:
+            bad_occurences += 1
+    return bad_occurences * 2
+    
+def period_in_df(meal):
+    url = (
+        'http://www.camara.leg.br/SitCamaraWS/sessoesreunioes.asmx/ListarDiscursosPlenario?'
+        'dataIni={}&dataFim={}&codigoSessao=&parteNomeParlamentar={}&siglaPartido=&siglaUF={}'
+    ).format(
+        meal.issue_date.strftime('%d/%m/%Y'),
+        meal.issue_date.strftime('%d/%m/%Y'),
+        meal.congressperson_name,
+        meal.congressperson_state
+    )
+    return '<a href="{}" target="_blank">{} - {}</a>'.format(
+        url,
+        meal.first_speech_at.strftime('%H:%M:%S'),
+        meal.last_speech_at.strftime('%H:%M:%S')
+    )
+    
+suspects2 = meals_outside_df_during_speeches.copy()
+suspects2.timestamps.fillna('', inplace=True)
+suspects2.receipt_dates.fillna('', inplace=True)
+suspects2['score'] = suspects2.apply(score, axis='columns')
+suspects2['period_in_df'] = suspects2.apply(period_in_df, axis='columns')
+suspects2 = suspects2.query('score > 0').sort_values('score', ascending=False)
+print("Suspicious reimbursements:", len(suspects2))
+
+suspects2 = suspects2[~suspects2.document_id.isin(suspects.document_id)]
+print("Suspicious reimbursements that were not found using presences info:", len(suspects2))
+
+
+# In[20]:
+
+report(suspects2.head(20), [
+    'document_id',
+    'receipt',
+    'issue_date', 
+    'congressperson_name',
+    'score',
+    'company_name',
+    'company_state', 
+    'total_net_value', 
+    'period_in_df'
+])
+
+
+# ## Meals in DF while the congressperson was outside DF
+# 
+# In the past we also found cases where the congressperson was on an official mission outside DF but requested a reimbursement for a meal in DF
+
+# In[21]:
+
+missions = pd.read_csv('../data/2017-06-05-official-missions.xz', dtype={'report_details_link': np.str}, low_memory=False)
+print("Total official mission records:", len(missions))
+
+missions = missions.query('canceled == "No"')
```

Comment:
This is the biggest change around official missions logic. While working on this I noticed that they introduced a new column indicating whether a mission was canceled. After filtering out the cancelled ones the amount of suspicious reimbursements reduced a lot.

/cc @rodolfo-viana